src/main/java/net/neoforged/neoforge/fluids/crafting/UnitAmountFluidIngredient.java

```diff
@@ -0,0 +1,267 @@
+     */
+    public static UnitAmountFluidIngredient of(TagKey<Fluid> tag, int amount, FluidUnit unit) {
+        return new UnitAmountFluidIngredient(FluidIngredient.tag(tag), List.of(Map.of(unit, amount)));
+    }
+
+    /**
+     * Factory method for creating any ingredient.
+     */
+    @SafeVarargs
+    public static UnitAmountFluidIngredient of(FluidIngredient ingredient, Map<FluidUnit, Integer>... amounts) {
+        if (amounts.length == 0)
+            return new UnitAmountFluidIngredient(ingredient, DEFAULT_AMOUNT);
+        else
+            return new UnitAmountFluidIngredient(ingredient, Arrays.stream(amounts)
+                    .map(UnitAmountFluidIngredient::validateNotEmpty)
+                    .map(DataResult::getOrThrow).toList());
+    }
+
+    private final FluidIngredient ingredient;
+    private final List<Map<FluidUnit, Integer>> amounts;
+
+    @Nullable
+    private FluidStack[] cachedStacks;
+
+    private UnitAmountFluidIngredient(FluidIngredient ingredient, List<Map<FluidUnit, Integer>> amounts) {
+        this.ingredient = ingredient;
+        this.amounts = amounts;
+    }
+
+    public FluidIngredient ingredient() {
+        return ingredient;
+    }
+
+    public List<Map<FluidUnit, Integer>> amounts() {
+        return amounts;
+    }
+
+    public Optional<Integer> amount(FluidStack stack) {
+        FluidUnitSpec fluidUnits = stack.getFluidType().getUnits();
+        for (Map<FluidUnit, Integer> amount : amounts) {
+            int accumulator = 0;
+            boolean valid = true;
+            for (Map.Entry<FluidUnit, Integer> entry : amount.entrySet()) {
+                Integer value = fluidUnits.getValue(entry.getKey());
+                if (value == null)
+                    valid = false;
+                else
+                    accumulator += value * entry.getValue();
+            }
+            if (valid)
+                return Optional.of(accumulator);
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Performs a size-sensitive test on the given stack.
+     *
+     * @return {@code true} if the stack matches the ingredient and has at least the required amount.
+     */
+    public boolean test(FluidStack stack) {
+        if (ingredient.test(stack)) {
+            return amount(stack).map(value -> stack.getAmount() >= value).orElse(false);
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Returns a list of the stacks from this {@link #ingredient}, with an updated {@link #amount}.
+     *
+     * @implNote the array is cached and should not be modified, just like {@link FluidIngredient#getStacks()}}.
+     */
+    public FluidStack[] getFluids() {
+        if (cachedStacks == null) {
+            cachedStacks = Stream.of(ingredient.getStacks())
+                    .map(input -> amount(input).map(input::copyWithAmount))
+                    .flatMap(Optional::stream)
```

Comment:
Nitpick, but these could easily be 1 map call and save a bit of effort, `.flatMap(input -> amount(input).map(input::copyWithAmount).stream())`